language/diem-vm/parallel-executor/src/outcome_array.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    errors::{Error, Result},
+    task::ExecutionStatus,
+};
+use once_cell::sync::OnceCell;
+
+pub(crate) struct OutcomeArray<T, E> {
+    // Hold the execution results for each individual transactions. Each cell should be set exactly
+    // once.
+    results: Vec<OnceCell<ExecutionStatus<T, Error<E>>>>,
```

Comment:
Yes. The reason why they are all wrapped with `OnceCell` is because we want this array to be shared by multiple threads, and in our design the execution status of each transaction should be setexactly once.

Reply:
The only difference between an `Option` is we can do interior mutability with `Sync`

Reply:
We actually have a guarantee that it's going to be a single write ever, so OnceCell mainly makes compiler happy and has the barrier, because actually OnceCell works if you tried many writers, it would just let only one succeed.